ambry-tools/src/main/java/com.github.ambry/store/DumpDataTool.java

```diff
@@ -179,39 +193,50 @@ private void compareIndexEntriesToLogContent(File indexFile, boolean checkLogEnd
-          long originalOffset = value.getOriginalMessageOffset();
-          if (originalOffset != -1) {
-            if (!coveredRanges.containsKey(originalOffset)) {
-              if (startOffset.getOffset() > originalOffset) {
-                logger.trace("Put Record at {} with delete msg offset {} ignored because it is prior to startOffset {}",
-                    originalOffset, value.getOffset(), startOffset);
-              } else {
-                DumpDataHelper.LogBlobRecordInfo logBlobRecordInfo =
-                    DumpDataHelper.readSingleRecordFromLog(randomAccessFile, originalOffset, clusterMap);
-                coveredRanges.put(originalOffset, originalOffset + logBlobRecordInfo.totalRecordSize);
-                logger.trace("PUT Record {} with start offset {} and end offset {} for a delete msg {} at offset {} ",
-                    logBlobRecordInfo.blobId, originalOffset, (originalOffset + logBlobRecordInfo.totalRecordSize),
-                    key.getID(), value.getOffset());
-                if (!logBlobRecordInfo.blobId.getID().equals(key.getID())) {
-                  logger.error("BlobId value mismatch between delete record " + key.getID() + " and put record "
-                      + logBlobRecordInfo.blobId.getID());
+      if (value.getOffset().getOffset() < logFileSize) {
+        boolean success = readFromLogAndVerify(randomAccessFile, key.getID(), value, coveredRanges);
+        if (success) {
+          if (isDeleted) {
+            long originalOffset = value.getOriginalMessageOffset();
+            if (originalOffset != -1) {
+              if (!coveredRanges.containsKey(originalOffset)) {
+                if (startOffset.getOffset() > originalOffset) {
+                  logger.trace(
+                      "Put Record at {} with delete msg offset {} ignored because it is prior to startOffset {}",
+                      originalOffset, value.getOffset(), startOffset);
+                } else {
+                  DumpDataHelper.LogBlobRecordInfo logBlobRecordInfo =
+                      DumpDataHelper.readSingleRecordFromLog(randomAccessFile, originalOffset, clusterMap);
+                  coveredRanges.put(originalOffset, originalOffset + logBlobRecordInfo.totalRecordSize);
+                  logger.trace("PUT Record {} with start offset {} and end offset {} for a delete msg {} at offset {} ",
+                      logBlobRecordInfo.blobId, originalOffset, (originalOffset + logBlobRecordInfo.totalRecordSize),
+                      key.getID(), value.getOffset());
+                  if (!logBlobRecordInfo.blobId.getID().equals(key.getID())) {
+                    logger.error("BlobId value mismatch between delete record " + key.getID() + " and put record "
+                        + logBlobRecordInfo.blobId.getID());
+                  }
                 }
               }
             }
           }
+        } else {
+          logger.error("Failed for key {} with value {} ", key, value);
```

Comment:
readFromLogAndVerify() would have printed the error details. This is just to print the final status incase the blob deserialization or comparison failed.